xmpp/auth.py

```diff
@@ -68,7 +68,7 @@ def plugin(self,owner):
 
         if query.getTag('digest'):
             self.DEBUG("Performing digest authentication",'ok')
-            query.setTagData('digest',sha1(owner.Dispatcher.Stream._document_attrs['id']+self.password).hexdigest())
+            query.setTagData('digest',sha1((owner.Dispatcher.Stream._document_attrs['id']+self.password).encode('utf-8')).hexdigest())
```

Comment:
With a wrapper function like
```python
def HDSHA1(some):
    if IS_PYTHON3 and isinstance(some, str):
        some = some.encode("utf-8")
    return sha1(some).hexdigest()
```

that part might look like

```suggestion
            query.setTagData('digest',HDSHA1(owner.Dispatcher.Stream._document_attrs['id']+self.password))
```

Reply:
What would you think about adopting the `six` helper functions instead (assuming that adding the dependency on the `six` library is acceptable)? `six.ensure_str()` and `six.ensure_binary()` should be better than the custom PYTHON3 conditionals. I think it would be clearer to write the code without additional custom helper functions where possible (and the SASL helpers could probably be made local to the `SASLHandler` function so that they're close by where they're used).

I suspect that the reason the wrapper functions are called `H` (hash), `HH` (hex-hash), and `C` (concatenate?) are because that's the notation used in [rfc2831 Section 2.1.2.1](https://datatracker.ietf.org/doc/html/rfc2831#section-2.1.2.1). NonSASL authentication is described in [xep-0078 Use Cases](https://xmpp.org/extensions/xep-0078.xml#usecases) and doesn't have the same sort of examples.

(side-note: 0k authentication should probably be removed. I couldn't find any documentation for it, and xep-0078 only briefly mentions it in section 2, and that it is unnecessary).

Reply:
Hi Norman. I believe introducing `six`, in order to streamline the corresponding bits in the code, is a good idea. I gave it a stab with #59.